[RFC] Use new-style arch_switch instead of arch_swap

carlocaione opened the pull request:
For now this is only a RFC to discuss the aarch64 switch to the new `arch_switch` API.

andrewboie commented on arch/arm/core/aarch64/switch.S:
```diff
@@ -36,6 +36,10 @@ GDATA(_k_neg_eagain)
+ *  - x16 <- new_thread->switch_handle
  */
 GTEXT(z_arm64_pendsv)
 SECTION_FUNC(TEXT, z_arm64_pendsv)
```
could we please re-name this stuff, since we are not actually doing the exception dance that aarch32 is doing?

carlocaione replied:
Yeah, it's time to rename that stuff. V2 coming.

stephanosio commented on arch/arm/core/aarch64/isr_wrapper.S:
```diff
@@ -83,15 +83,28 @@ SECTION_FUNC(TEXT, _isr_wrapper)
+	 * x17 <- old_thread->switch_handle
+	 */
+	mov	x16, x3
+	mov	x17, x2
```
We could `ldr` into `x16` and `x17` directly above:
```
	/* Check if we need to context switch */
	ldr	x16, [x1, #_kernel_offset_to_ready_q_cache]
	ldr	x17, [x1, #_kernel_offset_to_current]
	cmp	x16, x17
```

stephanosio commented on arch/arm/core/aarch64/isr_wrapper.S:
```diff
@@ -83,15 +83,28 @@ SECTION_FUNC(TEXT, _isr_wrapper)
+
+	/* Now we can fetch the thread to run from the ready queue cache */
+	ldr	x2, [x1, #_kernel_offset_to_ready_q_cache]
+	str	x2, [x1, #_kernel_offset_to_current]
```
Since `ready_q_cache` is already in `x16`, we could just do:
```
	str	x16, [x1, #_kernel_offset_to_current]
```

andyross replied:
Design-wise, this isn't really where you want to be.  You can't use _current at all in SMP (because you'd have to recompute it on the fly every time a CPU selected it to run, which defeats the point of the optimization).  The intent is that arch_switch() architectures call z_get_next_switch_handle() out of their interrupt exit path, and rely on the built-in z_swap() in the kernel which does the scheduler logic for you and just provides the switch handle and an output slot for the saved one.

carlocaione replied:
So IIUC since I'm using the thread pointer as `switch_handler` (see https://github.com/zephyrproject-rtos/zephyr/pull/22877/files#diff-f51234a0c31a6e7a874cac5d2595b0e9R84) I cannot pass the `_current` pointer directly to `z_get_next_switch_handle()`in the interrupt exit path?

carlocaione replied:
@andyross ^ as for my question above, is it allowed to directly pass `_current` to `z_get_next_switch_handle()`?

carlocaione replied:
Pushed a new revision

stephanosio commented on arch/arm/core/aarch64/switch.S:
```diff
@@ -23,34 +23,23 @@ GDATA(_kernel)
+	 * dummy_thread->switch_handle is 0 when setup by z_cstart()
+	 */
+	cmp	x17, #0
+	beq	null_switch
```
I wonder if there is a better way of handling this (i.e. something that does not require a `NULL` check whenever a context switch occurs)?

andrewboie replied:
@andyross might have some advice here

andyross replied:
The dummy thread should never be switched "to", so its switch handle isn't ever supposed to be inspected.  This sounds like something is working backwards here.   The idea behind arch_switch is that you don't muck with the scheduler internals.  The kernel decides for you what switch handle you should be switching to, and asks that you store the old one through the pointer in the second argument.

carlocaione commented:
@andyross I pushed a new revision, mostly as RFC. Can you please check whether I'm using the new API correctly?

andyross commented on arch/arm/core/aarch64/isr_wrapper.S:
```diff
@@ -97,12 +97,34 @@ SECTION_FUNC(TEXT, _isr_wrapper)
+	 * Save _current in x17 because this can be changed by
+	 * z_get_next_switch_handle()
+	 */
+	ldr	x17, [x1, #_kernel_offset_to_current]
```
This might be more complicated than necessary.  The intent of the "interrupted" argument to z_get_next_switch_handle() is that it's a switch handle constructed on interrupt **entry**[1] to contain the saved state of whatever got interrupted.   So once you have it (it doesn't have to be a thread pointer, but it works just fine if it is) you just leave it in a callee-save register across the ISR, then use it here.

Also a future note: in SMP contexts, there is no "_current" location in memory (because obviously every CPU has its own current thread).  The architecture is expected to store a current `struct _cpu` pointer (see `arch_curr_cpu()`) somewhere in the CPU outside the toolchain-visible register set (e.g. the FS segment register on x86, a MISC special register on xtensa, etc...) that it can be accessed easily, and this has a current field.

[1] Obviously there can be complexities introduced for optimization reasons.  For example you might save off only the caller-save registers on entry, and save the callee-save ones here only in the circumstance where the threads are different.  But conceptually, "interrupted" is something you should already have once the handler starts running.

carlocaione replied:
@andyross I'm reworking this PR but I still have one question.

In my current rework I'm using SP as switch handler as detailed in the `kernel_arch_interface.h` comment, and that's fine for the cooperative switching.

I'm a bit puzzled about the non-cooperative switching though. I'm using `z_get_next_switch_handle()` to get the handle of the next thread to be scheduled. Let's say that a new thread must be scheduled so I'm supposed to save the current context (basically my SP after pushing all the registers) into the `switch_handler` of the current task before this is switched out. How can I get access to the `switch_handler` without using the `_current` pointer? Should we go through `struct _cpu` as you pointed out in the previous message?

andyross commented on arch/arm/core/aarch64/switch.S:
```diff
@@ -33,18 +33,15 @@ GDATA(_k_neg_eagain)
-	ldr	x1, =_kernel
-	ldr	x2, [x1, #_kernel_offset_to_current]
+	/*
+	 * dummy_thread->switch_handle is 0 when setup by z_cstart()
```
This still seems like something that could be fixed without the null check.  Check the way x86_64 does it -- it doesn't load the switch_handle field from a thread, it offsets the switch_handle address it was passed to find the enclosing thread struct.

andyross commented on arch/arm/core/aarch64/switch.S:
```diff
@@ -222,6 +220,7 @@ offload:
 	b	inv
 
 context_switch:
+	ldr	x17, [x17]
```
Another SMP note: You're going to want to store through that switched_from pointer at the end of all context saving with some non-null value (and on ARM include a write barrier, IIRC).  See the comments (they're relatively new) at https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/include/kernel_arch_interface.h#L77

Basically this is used as a synchronization mechanism to be sure the thread being saved is in a valid state.

andyross reviewed (commented):
Some notes.  I think this looks broadly correct, though there are a few more things you'll need for SMP.

andrewboie commented:
Based on the technical discussions here, are there any obvious improvements we could make to arch_interface.h/kernel_arch_interface.h documentation ? no need to do in this PR

carlocaione commented:
I'm moving the cleanup commits to #23638 leaving this story only for the `arch_swap` -> `arch_switch` changes.

ioannisg commented:
Rebase is needed now @carlocaione

carlocaione commented:
> Rebase is needed now @carlocaione

This is very much still WiP. I'll rebase when I have something ready for PR :)

carlocaione committed:
arm: AArch64: Save the callee-saved register directly into SP

In the current implementation in case of context switch we are saving
the callee-saved registers separately into a callee_saved struct in the
thread struct together with the SP pointer. This is redundant since we
can directly save these registers onto the stack and just save SP.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
arch: aarch64: Use new-style arch_switch instead of arch_swap

Even though SMP is not supported yet, the first step needed is to switch
to the _arch_switch() API that is required for an SMP-aware scheduler.

In this commit we use the SP as switch_handle.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione commented:
@andyross @stephanosio @ioannisg @Abhishek-brcm 
This is going to fail on shippable but as usual I want to keep this story as RFC because I still have some questions that I'd like to address and discuss with you.

In this PR there are 3 commits:
- 46472770 this is taken from #23685 only to satisfy dependencies
- 670bec59 in this commit we are reducing the callee_saved register to only SP. Naively I think this is a straightforward thing to do but I wonder why not so many architectures do that already. Is there any fundamental reason to have the list of `callee_saved` registers in a struct instead of saving them on SP and just store that?
- 3904c2b9 this is the actual commit to switch to `arch_switch`. Given the previous commit I'm directly using the SP as `switch_handle`. About this I have mainly two questions:

1. About the `_current` usage as written already here https://github.com/zephyrproject-rtos/zephyr/pull/22877#discussion_r397091570
2. Given that we are now storing SP into the `switch_handle` what's the reason to have the pointer to `callee_saved` still around?

stephanosio commented:
> * Is there any fundamental reason to have the list of `callee_saved` registers in a struct instead of saving them on SP and just store that?

While most of the current arch implementations can probably cope with storing everything in the thread stack, this can potentially become a hassle in the future if we want to implement more advanced context switching features.

For instance, if we want to implement lazy FP context restoration by restoring the FP registers on-demand after the first FP instruction is executed after a thread is switched in, we must have these registers in the thread struct so that they can be restored after the thread is switched in and the context stack frame is destroyed.

andyross commented:
There's no reason not to have a hybrid.  In particular lazy FP doesn't actually get touched in the proper context switch code anyway and doesn't have to share an implementation.  It could stay in the thread struct (or somewhere else, maybe at the base of the thread stack in an area reserved only for thread with a "I_WANT_FP" flag or something, etc...) while the GPRs live on the stack, etc...

carlocaione commented:
Now this is successful on Shippable as well.

ioannisg reviewed (commented):
@carlocaione i wonder about the first patch - saving the callee-saved directly onto the stack; why is this considered an enhancement? All threads need to preserve callee-saved on context switch, so having them on a thread-struct does not consumer more space of processing-time...

carlocaione commented:
> @carlocaione i wonder about the first patch - saving the callee-saved directly onto the stack; why is this considered an enhancement? All threads need to preserve callee-saved on context switch, so having them on a thread-struct does not consumer more space of processing-time...

The idea is pretty much taken from https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/include/kernel_arch_interface.h#L77 (see the 4 steps).

Saving all the context on the SP and using that as `switch_handle` means that when the `switch_handle` is passed down to the `z_arm64_context_switch()` routine from `arch_switch()` we can just use that to access the context of the new task.

carlescufi commented:
@carlocaione with feature freeze coming on the 8th of May, is this something you'd like to get in before then?

carlocaione commented:
> @carlocaione with feature freeze coming on the 8th of May, is this something you'd like to get in before then?

I don't think that's possible given that this is pretty much in a RFC state.

nashif commented:
@carlocaione  can we close this?

carlocaione commented:
> @carlocaione can we close this?

Yes. Superseded by #27766
